scraper/src/libretexts2zim/generator.py

```diff
@@ -34,39 +39,101 @@ class MissingDocumentError(Exception):
-            "--shelves-include",
-            help="Includes only shelves matching the given regular expression.",
+            "--page-title-include",
+            help="Includes only pages with title matching the given regular "
+            "expression, and their parent pages for proper navigation. Can be combined"
+            " with --page-id-include (pages with matching title or id will be included"
+            ")",
             metavar="REGEX",
         )
 
         parser.add_argument(
-            "--shelves-exclude",
-            help="Excludes shelves matching the given regular expression.",
+            "--page-id-include",
+            help="CSV value of page ids to include. Parent pages will be included as "
+            "well for proper navigation. Can be combined with --page-title-include "
+            "(pages with matching title or id will be included)",
+        )
+
+        parser.add_argument(
+            "--page-title-exclude",
+            help="Excludes pages with title matching the given regular expression",
             metavar="REGEX",
         )
 
+        parser.add_argument(
+            "--root-page-id",
+            help="ID of the root page to include in ZIM. Only this page and its"
+            " subpages will be included in the ZIM",
+        )
+
     @staticmethod
     def of(namespace: argparse.Namespace) -> "ContentFilter":
         """Parses a namespace to create a new DocFilter."""
         return ContentFilter.model_validate(namespace, from_attributes=True)
 
-    # TODO: implement filtering of shelves based on configured regex
-    # def filter(self, shelves: list[LibretextsShelve]) -> list[LibretextsShelve]:
-    #     """Filters docs based on the user's choices."""
-    #     selected: list[LibretextsShelve] = []
-    #     for shelve in shelves:
-    #       ....
-    #     return selected
+    def filter(self, page_tree: DekiTree) -> list[DekiPage]:
+        """Filters pages based on the user's choices."""
+
+        if self.root_page_id:
+            page_tree = page_tree.sub_tree(self.root_page_id)
+
+        title_include_re = (
+            re.compile(self.page_title_include, re.IGNORECASE)
+            if self.page_title_include
+            else None
+        )
+        title_exclude_re = (
+            re.compile(self.page_title_exclude, re.IGNORECASE)
+            if self.page_title_exclude
+            else None
+        )
+        id_include = (
+            [page_id.strip() for page_id in self.page_id_include.split(",")]
+            if self.page_id_include
+            else None
+        )
+
+        def is_selected(
+            title_include_re: re.Pattern[str] | None,
+            title_exclude_re: re.Pattern[str] | None,
+            id_include: list[str] | None,
+            page: DekiPage,
+        ) -> bool:
+            return (
+                (
+                    not title_include_re
+                    or title_include_re.search(page.title) is not None
```

Comment:
I think the `is not None` is not required here ; If confirmed, I'd remove it as this block is already intense to read

Reply:
Type checker is not happy, even if I agree it would probably work.